Dockerfile

```diff
@@ -1,61 +1,58 @@
-    apt-get install -y pkg-config openssl libssl-dev libtag1-dev libtagc0-dev curl yasm build-essential wget libbz2-dev zlib1g-dev &&\
-    curl -sL https://deb.nodesource.com/setup_8.x | bash - &&\
-    apt-get install -y nodejs 
-
-COPY . /audioserve_src
-
-RUN curl https://sh.rustup.rs -sSf | sh -s -- -y
-
-RUN mkdir ffmpeg-static &&\
-    curl -sL https://johnvansickle.com/ffmpeg/releases/ffmpeg-release-amd64-static.tar.xz | tar xJv -C ffmpeg-static --strip-components=1 --wildcards "*/ffmpeg" &&\
-    cp /ffmpeg-static/ffmpeg /usr/bin
+RUN apt-get update && \
+    apt-get install -y \
+    curl \
+    libbz2-dev \
+    libssl1.0-dev \
+    libtag1-dev \
+    libtagc0-dev \
+    openssl \
+    pkg-config \
+    yasm && \
+    mkdir ffmpeg-static && \
+    curl -sL https://johnvansickle.com/ffmpeg/releases/ffmpeg-release-amd64-static.tar.xz | \
+    tar xJv -C ffmpeg-static --strip-components=1 --wildcards "*/ffmpeg" && \
+    mv /ffmpeg-static/ffmpeg /usr/bin/ffmpeg
+
+WORKDIR /src
+COPY Cargo.toml Cargo.lock /src/
+# To cache this step, we only copy the toml file above
+RUN cargo fetch
+
+COPY . /src
+
+RUN cargo build --release ${FEATURES} && \
+    cargo test --release ${FEATURES}
 
+FROM node:8-alpine as frontend
 
-RUN export PATH=${HOME}/.cargo/bin:$PATH &&\
-    cd audioserve_src &&\
-    cargo build --release ${FEATURES} &&\
-    cargo test --release ${FEATURES}
+WORKDIR /src
+COPY client /src
 
-RUN cd audioserve_src/client &&\
-    npm install &&\
+RUN npm install && \
     npm run build
 
-RUN mkdir /ssl &&\
-    cd /ssl &&\
-    openssl req -newkey rsa:2048 -nodes -keyout key.pem -x509 -days 365 -out certificate.pem \
-        -subj "/C=CZ/ST=Prague/L=Prague/O=Ivan/CN=audioserve" &&\
-    openssl pkcs12 -inkey key.pem -in certificate.pem -export  -passout pass:mypass -out audioserve.p12 
-
 FROM debian:stretch-slim
 
 VOLUME /audiobooks
-COPY --from=build /audioserve_src/target/release/audioserve /audioserve/audioserve
-COPY --from=build /audioserve_src/client/dist /audioserve/client/dist
-COPY --from=build /ssl/audioserve.p12 /audioserve/ssl/audioserve.p12
+COPY --from=build /src/target/release/audioserve /audioserve/audioserve
+COPY --from=frontend /src/client/dist /audioserve/client/dist
 COPY --from=build /ffmpeg-static/ffmpeg /usr/bin
 
-RUN adduser audioserve &&\
-    chown -R audioserve:audioserve /audioserve &&\
-    apt -o Acquire::https::No-Cache=True -o Acquire::http::No-Cache=True update &&\
+RUN adduser audioserve && \
+    chown -R audioserve:audioserve /audioserve && \
+    apt -o Acquire::https::No-Cache=True -o Acquire::http::No-Cache=True update && \
     apt-get install -y libssl1.1 libtag1v5 libtagc0 libbz2-1.0
-   
+
 WORKDIR /audioserve
 USER audioserve
-
-ENV SECRET=mypass
-ENV SSLKEY=./ssl/audioserve.p12
-ENV SSLPASS=mypass
-ENV DIRS=/audiobooks
 ENV PORT=3000
 
 EXPOSE ${PORT}
 
-COPY audioserve.sh .
-
-CMD ./audioserve.sh
+ENTRYPOINT ["/audioserve/audioserve"]
```

Comment:
I'd like to keep wrapper script - for demo/testing purposes it's invaluable as one can play with command line args. $PORT is for instance needed if one wants to try on heroku.

Reply:
Docker allows passing command line args to the entrypoint, [I was relying on that](https://stackoverflow.com/a/40312311)

Will keep `PORT`, but it might be better handled at the app itself. (instead of reading it as env and passing it as a argument)

Reply:
I see now - that could be easier solution then script unless we would need to mess with args - like $PORT or $USE_UNSECURE_CERT